### PR TITLE
refactor(lodash): simplify `clamp` and `debounce`

### PR DESCRIPTION
**Pull request by sadeghbarati**
## Description

Remove `lodash` utilities
Simple utils without additional type checking

<del>Refactored `debounce` is copy version of lodash, with fewer type checks & other internal lodash utilities
is debounce `cancel`, `flush` or `isPending` needed to be included in this package?</del>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/utils/debounce.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  cancel: () => void;
+  flush: () => void;
+  isPending: () => boolean;
+}
+
+export interface DebouncedReturn<T extends (...args: any[]) => ReturnType<T>> extends IDebounceReturn {
+  (...args: Parameters<T>): ReturnType<T> | undefined;
+}
+
+export function debounce<T extends (...args: any) => ReturnType<T>>(fn: T, wait?: number, options?: DebounceOptions): DebouncedReturn<T> {
```
Comment (idiotWu) [inline src/utils/debounce.ts:17]: I don't think we need an implementation that covers every corner case. We just need a minimum implementation like:

```js
// we only used the `leading` option in our code
// NEED TEST!
function debounce(fn, wait, leading) {
  let timer;

  return (...args) => {
    clearTimeout(timer);
 
    let ms = leading ? 0 : wait;
    leading = false;

    timer = setTimeout(() => {
      fn(...args);
    }, ms);
  };
}
```

**Review by idiotWu — CHANGES_REQUESTED** [review]

File: src/utils/debounce.ts (hunk 1 above)
Comment (idiotWu) [reply src/utils/debounce.ts:17]: See also: https://css-tricks.com/debouncing-throttling-explained-examples/

Comment (sadeghbarati) [reply src/utils/debounce.ts:17]: lodash **`useRaf`** pushed me to copy same logic, cause most of `debounce` utils on web using similar variables and logics as lodash debounce

what is ur idea about refactor same code you shared with both `setTimeout` and `requestAnimationFrame`?

cause `requestAnimationFrame` is more efficient than `setTimeout`

Comment (sadeghbarati) [reply src/utils/debounce.ts:17]: GitHub Copilot debounce suggestion :alien:
```js
function debounce(func, wait, leading) {
  let timeout
  return function () {
    const context = this; const args = arguments
    const later = function () {
      timeout = null
      if (!leading)
        func.apply(context, args)
    }
    const callNow = leading && !timeout
    clearTimeout(timeout)
    timeout = setTimeout(later, wait)
    if (callNow)
      func.apply(context, args)
  }
}
```

Comment (idiotWu) [reply src/utils/debounce.ts:17]: Copilot's version is functionally the same as mine except that I forgot to support `this` context. 😆

Comment (idiotWu) [reply src/utils/debounce.ts:17]: Fixed version:

```js
// we only used the `leading` option in our code
// NEED MORE TESTS!
function debounce(fn, wait, leading) {
  let timer;
  let lastCalledAt = -Infinity;

  function handleLeading(thisArg, args) {
    const now = Date.now();
    const elapsed = now - lastCalledAt;

    lastCalledAt = now;

    if (elapsed >= wait) {
      fn.apply(thisArg, args);
    }
  }

  return function debouncedFn(...args) {
    if (leading) {
      handleLeading(this, args);
      return;
    }

    clearTimeout(timer);

    timer = setTimeout(() => {
      fn.apply(this, args);
    }, wait);
  };
}

```

**Comment by sadeghbarati** [comment]
what's your idea about `requestAnimationFrame`, do you think `setTimeout` is enough? :v:

File: src/utils/debounce.ts [hunk 2]
```diff
@@ -1,146 +1,29 @@
...
-    }
-    return setTimeout(pendingFunc, wait);
-  }
 
-  function cancelTimer(id) {
-    useRAF ? window.cancelAnimationFrame(id) : clearTimeout(id);
-  }
+export function debounce(fn: Function, wait?: number, leading?: boolean) {
+  let timer;
+  let lastCalledAt = -Infinity;
```
Comment (sadeghbarati) [inline src/utils/debounce.ts:4]: Why not use `null` or `undefined`? 🤔

Comment (idiotWu) [reply src/utils/debounce.ts:4]: 1. it has the type `number`,
2. you can perform arithmetic operations when assigning it to `Infinity`.

File: src/utils/debounce.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    const elapsed = now - lastCalledAt;
+
+    lastCalledAt = now;
+
+    if (elapsed >= wait!) {
+      fn.apply(thisArg, args);
+    }
+  }
+
+  return function debouncedFn(this: unknown, ...args) {
```
Comment (idiotWu) [inline src/utils/debounce.ts:17]:
```ts
return function debouncedFn(this: unknown, ...args: Parameters<T>)
```

File: src/utils/debounce.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+
+export function debounce(fn: Function, wait?: number, leading?: boolean) {
+  let timer;
+  let lastCalledAt = -Infinity;
+
+  function handleLeading(thisArg, args) {
```
Comment (idiotWu) [inline src/utils/debounce.ts:6]:
```ts
function handleLeading(thisArg: unknown, args: Parameters<T>)
```

Maybe it's better to move this function into the returned `debouncedFn` block?

**Review by idiotWu — CHANGES_REQUESTED** [review]

File: src/utils/debounce.ts (hunk 4 above)
Comment (sadeghbarati) [reply src/utils/debounce.ts]: agreed, prevent duplication

File: src/utils/debounce.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+
+export function debounce<T extends (...args: any[]) => void>(fn: T, wait?: number, leading?: boolean) {
+  let timer: number | NodeJS.Timeout;
+  let lastCalledAt = -Infinity;
+
```
Comment (idiotWu) [inline src/utils/debounce.ts:5]: Remove the blank line.

File: src/utils/debounce.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+export function debounce<T extends (...args: any[]) => void>(fn: T, wait?: number, leading?: boolean) {
+  let timer: number | NodeJS.Timeout;
```
Comment (idiotWu) [inline src/utils/debounce.ts:3]: I don't think our environment includes node.js. Any reason you add `NodeJS.Timeout` here? 🤔

File: src/utils/debounce.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+
```
Comment (idiotWu) [inline src/utils/debounce.ts:1]: Remove the blank line.

File: src/utils/debounce.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    if (leading) {
+      const now = Date.now();
+      const elapsed = now - lastCalledAt;
+
+      lastCalledAt = now;
+
+      if (elapsed >= wait!) {
+        fn.apply(this, args);
+      }
+      return;
```
Comment (idiotWu) [inline src/utils/debounce.ts:17]: Maybe it's better to use an `else` block here because this `if` block is quite long.

File: src/utils/debounce.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+
+export function debounce<T extends (...args: any[]) => void>(fn: T, wait?: number, leading?: boolean) {
```
Comment (idiotWu) [inline src/utils/debounce.ts:2]: We should make `wait` a required parameter here, i.e. `wait: number`. (Then you should also remove the non-null assertions below: `wait!` -> `wait`.)

**Review by idiotWu — CHANGES_REQUESTED** [review]

File: src/utils/debounce.ts (hunk 9 above)
Comment (idiotWu) [reply src/utils/debounce.ts:2]: Or you can set a default value for it like `wait: number = 0`.

File: src/utils/debounce.ts (hunk 7 above)
Comment (sadeghbarati) [reply src/utils/debounce.ts:1]: sry, I got used to tools that clean or lint everything :dizzy_face:

File: src/utils/debounce.ts (hunk 9 above)
Comment (sadeghbarati) [reply src/utils/debounce.ts]: Thanks for tips ❤️‍🔥

File: src/utils/debounce.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+export function debounce<T extends (...args: any[]) => void>(fn: T, wait: number = 0, leading?: boolean) {
+  let timer: number | ReturnType<typeof setTimeout>;
```
Comment (idiotWu) [inline src/utils/debounce.ts:2]: Why is `ReturnType<typeof setTimeout>` needed here?

Comment (sadeghbarati) [reply src/utils/debounce.ts:2]: I'm using takeover mode `Volar` instead of Vscode builtin Typescript
I'll check with Vscode Typescript


```ts
let timer: number; // only type number will cause typescript error

// Type 'Timeout' is not assignable to type 'number'.
timer = setTimeout(() => {
  fn.apply(this, args);
}, wait);
```

---

```ts
// ReturnType<typeof setTimeout> => number

let timer: number | ReturnType<typeof setTimeout>;

// no error
timer = setTimeout(() => {
  fn.apply(this, args);
}, wait);
```

Comment (sadeghbarati) [reply src/utils/debounce.ts]: Ow, you're right :sweat:👍 
This is a client-side package

**Review by idiotWu — APPROVED** [review]
LGTM!
I believe we should add e2e tests asap. 😂
